8.0 fix old style computed fields dle

gurneyalex opened the pull request:
This is a port of https://github.com/odoo/odoo/pull/6315 to OCB with my additional (yet unmerged) PR https://github.com/odoo-dev/odoo/pull/106 included (adding an automated test). 

It fixes the following bug: when records are created through a workflow chain, some stored function fields may not be computed at record creation. This happens for stock.move in the prepaid sale order flow. 

See #6053 for the discussion.

lmignon commented:
@gurneyalex Thank you for the tests!!! At least for this bug we are now sure that no regression will occur in the future.
I'm a little surprised to see that part of the fix is to go back to the old API ...

gurneyalex commented:
@lmignon the "going back to new API" is a revert of an ad-hoc fix (see commit message) which only fixed the stock.move product_qty field, and it is not clear to anyone why changing that field to the new API solved anything at that time. It was committed in 8.0 in some rush because not having product_qty on stock_move completely blocked the flow, but the bug potentially affects many other field...

gurneyalex commented:
A good way of testing if your instances are affected by this bug is to go through the stored function fields in the database and look for unexepected NULL values. You should have none of these for numeric fields or character fields (they get a value of 0 or '' respectively). A NULL in there means that the compute function was not called at all.

lmignon commented:
@gurneyalex The first commit should never have happened. Once again thank you for your tests.

pedrobaeza commented on addons/stock/stock.py:
```diff
@@ -1591,7 +1590,7 @@ def action_traceability(self, cr, uid, ids, context=None):
 # ----------------------------------------------------
 
-class stock_move(models.Model):
+class stock_move(osv.osv):
```
Why replace this?

gurneyalex replied:
see my answer to lmignon's question

pedrobaeza replied:
Ah, OK, curious that this was introduce upstream..

pedrobaeza commented:
Seems OK for me :+1:

guewen commented:
:+1:

lmignon commented:
:+1:

beledouxdenis committed:
[FIX] models: recompute nested old-style computed field

The `set` method of the one2many class returns a list
of the fields that require recomputation,
the computing of the function fields being delayed
for performances reasons.

Nevertheless, if the `set` method was called
through another `set` method, in other words,
nested `set` calls, the fields to recompute returned
by the second, nested, call to set were never recomputed,
the result list were simply lost.

e.g.:
```
create/write
│set
└─── create/write with recs.env.recompute set to False
    │set
        └─── create
             with recs.env.recompute set to False
```

To overcome this problem, the list of old api style
compute fields to recompute is stored
within the environment, and this list is cleared
each time the store_set_value has done its job of
recomputing all old api style compute fields.

opw-629650
opw-632624
closes #6053

beledouxdenis committed:
Revert "[FIX] stock: `product_qty` field is not recomputed at record creation."

This reverts commit a1da6c21325ae7121aaba643aa1eea6401649dec.

gurneyalex committed:
[ADD] sale, delivery: tests for sales prepaid flow

Test uses cases for revision 073e0244de326a7ba1e33493d7069e345e48d4e1

gurneyalex commented:
upstream PR updated -> I'll update this one.

gurneyalex commented:
Upstream PR merged. Closing.
